src/Common/src/System/Collections/Generic/EnumerableHelpers.Linq.cs

```diff
@@ -0,0 +1,53 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Collections.Generic
+{
+    /// <summary>Internal helper functions for working with enumerables.</summary>
+    internal static partial class EnumerableHelpers
+    {
+        /// <summary>Gets the count of the specified enumerable.</summary>
+        /// <param name="source">The enumerable to count.</param>
+        /// <returns>The count of the enumerable.</returns>
+        internal static int GetCount<T>(IEnumerable<T> source)
+        {
+            Debug.Assert(source != null);
+
+            var collectionOfT = source as ICollection<T>;
+            if (collectionOfT != null)
+            {
+                return collectionOfT.Count;
+            }
+            
+            var collection = source as ICollection;
+            if (collection != null)
+            {
+                return collection.Count;
+            }
+
+            return GetCountCore(source);
+        }
+
+        private static int GetCountCore<T>(IEnumerable<T> source)
```

Comment:
Self-note: We might even want to just make this method non-generic (use the untyped `IEnumerable` / `IEnumerator` interfaces), which means that this will only be jitted once. At that point we could move the non-generic ICollection typecast into here, and then this would be another helper method capable of getting the count of a non-generic enumerable.